ArcGISRuntimeSDKQt_CppSamples/Maps/DownloadPreplannedMap/DownloadPreplannedMap.cpp

```diff
@@ -0,0 +1,226 @@
+}
+
+DownloadPreplannedMap::~DownloadPreplannedMap() = default;
+
+void DownloadPreplannedMap::init()
+{
+  // Register the map view for QML
+  qmlRegisterType<MapQuickView>("Esri.Samples", 1, 0, "MapView");
+  qmlRegisterType<DownloadPreplannedMap>("Esri.Samples", 1, 0, "DownloadPreplannedMapSample");
+}
+
+MapQuickView* DownloadPreplannedMap::mapView() const
+{
+  return m_mapView;
+}
+
+// Set the view (created in QML)
+void DownloadPreplannedMap::setMapView(MapQuickView* mapView)
+{
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+
+  emit mapViewChanged();
+}
+
+void DownloadPreplannedMap::downloadMapArea(int index)
+{
+  if (!m_offlineMapTask)
+    return;
+
+  if (m_offlineMapTask->preplannedMapAreaList()->isEmpty())
+    return;
+
+  if (m_offlineMapTask->preplannedMapAreaList()->at(index)->loadStatus() != LoadStatus::Loaded)
+    return;
+
+  m_busy = true;
+  emit busyChanged();
+
+  QString path = QString(m_tempPath.path() + "/" + m_offlineMapTask->preplannedMapAreaList()->at(index)->portalItem()->title() + ".mmpk");
+
+  QDir tempDir(path);
+
+  if (m_mapExists)
+  {
+    if (m_mmpk)
+    {
+      delete m_mmpk;
+      m_mmpk = nullptr;
+    }
+
+    m_mmpk = new MobileMapPackage(path, this);
+
+    connect(m_mmpk, &MobileMapPackage::doneLoading, this, [this] (Error e)
+    {
+      if (!e.isEmpty())
+        qDebug() << e.message() << " - " << e.additionalMessage();
+
+      m_mapView->setMap(m_mmpk->maps().at(0));
+      m_busy = false;
+      emit busyChanged();
+      m_graphicsOverlay->setVisible(false);
+      m_viewingOnlineMaps = false;
+      emit viewingOnlineMapsChanged();
+    });
+
+    m_mmpk->load();
+    return;
+  }
+  else
+  {
+    connect(m_offlineMapTask, &OfflineMapTask::createDefaultDownloadPreplannedOfflineMapParametersCompleted, this, [this, path] (QUuid ,const DownloadPreplannedOfflineMapParameters &parameters)
+    {
+
+      m_params = DownloadPreplannedOfflineMapParameters(parameters);
+      m_params.setUpdateMode(PreplannedUpdateMode::NoUpdates);
+
+      m_preplannedMapJob = m_offlineMapTask->downloadPreplannedOfflineMap(m_params, path);
+      connect(m_preplannedMapJob, &DownloadPreplannedOfflineMapJob::jobDone, this, &DownloadPreplannedMap::onDownloadPreplannedMapJobCompleted);
+      m_preplannedMapJob->start();
+    });
+
+    m_offlineMapTask->createDefaultDownloadPreplannedOfflineMapParameters(m_offlineMapTask->preplannedMapAreaList()->at(index));
+  }
+}
+
+void DownloadPreplannedMap::checkIfMapExists(const int index)
+{
+  if (!m_offlineMapTask)
+    return;
+
+  if (m_offlineMapTask->preplannedMapAreaList()->isEmpty())
+    return;
+
+  if (m_offlineMapTask->preplannedMapAreaList()->at(index)->loadStatus() != LoadStatus::Loaded)
+    return;
+
+  QString path = QString(m_tempPath.path() + "/" + m_offlineMapTask->preplannedMapAreaList()->at(index)->portalItem()->title() + ".mmpk");
+  QDir tempDir(path);
+
+  if (tempDir.exists())
```

Comment:
I think you could simplify this logic to something like:
```suggestion
m_mapExists = tempDir.exists();
emit mapExistsChanged();
```